openstack/path_static_creds.go

```diff
@@ -0,0 +1,215 @@
+package openstack
+
+import (
+	"context"
+	"fmt"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/tokens"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/users"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+const (
+	pathStaticCreds = "static-creds"
+
+	staticCredsHelpSyn  = "Manage the Openstack static credentials with static roles."
+	staticCredsHelpDesc = `
+This path allows you to read OpenStack secret stored by predefined static roles.
+`
+)
+
+func (b *backend) pathStaticCreds() *framework.Path {
+	return &framework.Path{
+		Pattern: fmt.Sprintf("%s/%s", pathStaticCreds, framework.GenericNameRegex("role")),
+		Fields: map[string]*framework.FieldSchema{
+			"role": {
+				Type:        framework.TypeString,
+				Description: "Name of the role.",
+				Required:    true,
+			},
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.ReadOperation: &framework.PathOperation{
+				Callback: b.pathStaticCredsRead,
+			},
+		},
+		HelpSynopsis:    staticCredsHelpSyn,
+		HelpDescription: staticCredsHelpDesc,
+	}
+}
+
+func (b *backend) pathStaticCredsRead(ctx context.Context, r *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	roleName := d.Get("role").(string)
+	role, err := getStaticRoleByName(ctx, roleName, r)
+	if err != nil {
+		return nil, err
+	}
+
+	sharedCloud := b.getSharedCloud(role.Cloud)
+	cloudConfig, err := sharedCloud.getCloudConfig(ctx, r.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := sharedCloud.getClient(ctx, r.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	var data map[string]interface{}
+	switch r := role.SecretType; r {
+	case SecretToken:
+		tokenOpts := &tokens.AuthOptions{
+			Username: role.Username,
+			Password: role.Secret,
+			DomainID: role.DomainID,
+			Scope:    getScopeFromStaticRole(role),
+		}
+
+		token, err := createToken(client, tokenOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		authResponse := &authStaticResponseData{
+			AuthURL:  cloudConfig.AuthURL,
+			Username: role.Username,
+			Token:    token.ID,
+			DomainID: role.DomainID,
+		}
+
+		data = map[string]interface{}{
+			"auth": formStaticAuthResponse(
+				role,
+				authResponse,
+			),
+			"auth_type": "token",
+		}
+
+	case SecretPassword:
+		authResponse := &authStaticResponseData{
+			AuthURL:  cloudConfig.AuthURL,
+			Username: role.Username,
+			Password: role.Secret,
+			DomainID: role.DomainID,
+		}
+		data = map[string]interface{}{
+			"auth": formStaticAuthResponse(
+				role,
+				authResponse,
+			),
+			"auth_type": "password",
+		}
```

Comment:
Maybe place outer map generation to `formStaticAuthResponse` too to make it `formStaticAuthResponse(role, authResponse, authType)`?

Reply:
This `data` part forms yaml response which I think is more understandable when you can see how both `auth` and `auth_type` are formed. If you insist I will change this part.